Generic help for QGIS

alexbruy opened the pull request:
This is my attempt to implement generic help system as described in qgis/QGIS-Enhancement-Proposals#51.

Added new QGIS settings — location of the help files. User can specify as many sources as he wants/needs (local directories, remote or local web-servers) and prioritize them by moving entries up and down

![help-setting](https://cloud.githubusercontent.com/assets/776954/21506254/4cb6c0d8-cc77-11e6-9c84-1e00f4fde1c7.png)

Directory layout in those locations should be the same as we use at our documentation server, e.g. `$HOME/qgis-help/X.XX/LANG/docs/user_manual`, where `X.XX` — is QGIS version in dotted notation (major.minor) and `LANG` — locale.

As a help "key" relative path to the corresponding file should be used, e.g. `processing_algs/qgis/vector_general_tools.html#delete-duplicate-geometries`. QGIS will automatically construct full URL taking into account current QGIS version and locale. If QGIS UI locale differ from the system locale (locale override in QGIS settings enabled), QGIS locale will be used.

When help requested, QGIS will iterate over all locations one by one and check if it is accessible by issuing HTTP HEAD request for network resources and testing file existence for local directories. First accessible location will be used, all other locations won't be checked. If no accessible help found, build-in error page will be opened.

New `QgsHelp` class provides two methods:
- `openHelp()` to open help topic in the system default browser
- `helpUrl()` which returns QUrl, pointing to the help topic (can be used for displaying in custom dialogs)

alexbruy committed:
add possibility to configure location of the QGIS help

alexbruy committed:
new QgsHelp class that uses QGIS User Guide as help source

alexbruy commented:
@m-kuhn, @rduivenvoorde please take a look at this

nyalldawson commented on doc/nohelp.html:
```diff
@@ -0,0 +1,19 @@
+<body>
+<center><a href="http://qgis.org"><img src="qgis-icon-60x60.png" border=0></a></center>
+<h1><center>Oops!</center></h1>
+<p>Sorry, help is not available. Probably you have no Internet connection,
```
What about something like:

```
<p>QGIS help is currently not available. This is likely because:</p>
<ul>
<li>No internet connection is available</li>
<li>The offline help files are not available or the location of these files is not properly configured</li>
<li>No help content is available for the requested topic</li>
</ul>
<p>Please check that the correct location of the offline help files is specified in the QGIS options dialog (Settings&nbsp;&rarr; Options&nbsp;&rarr;System).</p>
```

nyalldawson commented on src/app/qgsoptions.cpp:
```diff
@@ -270,6 +270,21 @@ QgsOptions::QgsOptions( QWidget *parent, Qt::WindowFlags fl )
   }
 
+  //locations of the QGIS help
+  myPaths = mSettings->value( QStringLiteral( "help/helpSearchPath" ), "" ).toString();
```
This is fragile (I realise it's copied/pasted from elsewhere, but easier to fix new settings then existing ones). A better approach:

    QStringList helpPathList = mSettings->value( QStringLiteral( "help/helpSearchPath" ) ).toStringList();
    Q_FOREACH( const QString& path, helpPathList )
    {
        QTreeWidgetItem* item = new QTreeWidgetItem();
        item->setText( 0, path );
        item->setFlags( Qt::ItemIsEnabled | Qt::ItemIsSelectable | Qt::ItemIsEditable );
        mHelpPathTreeWidget->addTopLevelItem( item );
    }

This also has the advantage of
1. avoiding the ugly "my" prefixed variables
2. no issues if someone has a | in their paths
3. no use of "" literal (`QString()` should be used instead)

nyalldawson commented on src/app/qgsoptions.cpp:
```diff
@@ -1118,6 +1133,22 @@ void QgsOptions::saveOptions()
+      myPaths += item->text( 0 );
+    }
+  }
+  mSettings->setValue( QStringLiteral( "help/helpSearchPath" ), myPaths );
```
Following the above comment, this would become:

    QStringList helpPaths;    
    for ( int i = 0; i < mHelpPathTreeWidget->topLevelItemCount(); ++i )
    {
      if ( QTreeWidgetItem* item = mHelpPathTreeWidget->topLevelItem( i ) )
      {
        helpPaths << item->text( 0 );
      }
    }
    mSettings->setValue( QStringLiteral( "help/helpSearchPath" ), helpPaths );

nyalldawson commented on src/app/qgsoptions.cpp:
```diff
@@ -1735,6 +1766,59 @@ void QgsOptions::on_mBtnRemovePluginPath_clicked()
+void QgsOptions::on_mBtnAddHelpPath_clicked()
+{
+  QTreeWidgetItem* item = new QTreeWidgetItem();
+  item->setText( 0, QLatin1String( "" ) );
```
"" -> QString()

nyalldawson commented on src/core/qgshelp.cpp:
```diff
@@ -0,0 +1,145 @@
+#include "qgis.h"
+#include "qgsapplication.h"
+
+QgsHelp *QgsHelp::sHelp = nullptr; // Singleton instance
```
We are avoiding adding new singletons to the code - see https://github.com/qgis/qgis3.0_api/issues/42

The correct approach here would be copying how the task manager/action scope registry/profiler are implemented in QgsApplication. See
https://github.com/qgis/QGIS/blob/master/src/core/qgsapplication.h#L371
https://github.com/qgis/QGIS/blob/master/src/core/qgsapplication.h#L510
https://github.com/qgis/QGIS/blob/master/src/core/qgsapplication.cpp#L112
https://github.com/qgis/QGIS/blob/master/src/core/qgsapplication.cpp#L247
https://github.com/qgis/QGIS/blob/master/src/core/qgsapplication.cpp#L1506

alexbruy replied:
I'm afraid, in this case we will get circular dependency, as `QgsHelp` includes `qgsapplication.h` and uses `QgsApplication::pkgDataPath()`. I will look if we can avoid it.

nyalldawson commented on src/core/qgshelp.cpp:
```diff
@@ -0,0 +1,145 @@
+  QSettings settings;
+  QUrl helpNotFound = QUrl::fromLocalFile( QgsApplication::pkgDataPath() + "/doc/nohelp.html" );
+
+  QString paths = settings.value( QStringLiteral( "help/helpSearchPath" ), "" ).toString();
```
QStringList paths = settings.value( QStringLiteral( "help/helpSearchPath" ) ).toStringList();

nyalldawson commented on src/core/qgshelp.cpp:
```diff
@@ -0,0 +1,145 @@
+  bool overrideLocale = settings.value( QStringLiteral( "locale/overrideFlag" ), false ).toBool();
+  if ( overrideLocale )
+  {
+    qgisLocale = settings.value( QStringLiteral( "locale/userLocale" ), "" ).toString();
```
"" -> QString()

nyalldawson commented on src/core/qgshelp.cpp:
```diff
@@ -0,0 +1,145 @@
+    }
+    else
+    {
+      if ( !QFileInfo( helpPath.mid( 0, helpPath.lastIndexOf( "#" ) ) ).exists() )
```
Use QFileInfo::exists(....). It's much faster then QFileInfo(...).exists().

nyalldawson commented on src/core/qgshelp.h:
```diff
@@ -0,0 +1,73 @@
+     * @param key key which identified help topic
+     * @note added in QGIS 3.0
+     */
+    static void openHelp( const QString& key );
```
Following from the earlier comment about not using a singleton here, these methods would become const members rather than static methods.

alexbruy committed:
improve built-in "no help found" page

alexbruy committed:
QgsHelp is not singletone

alexbruy committed:
add SIP bindings

alexbruy committed:
improve QSettings handling

nyalldawson commented:
Just a thought - given that there's no member variables in QgsHelp, why not just make all it's methods static and do away with the need to create an instance in the first place?

alexbruy committed:
make all QgsHelp methods static

m-kuhn commented on src/core/qgshelp.h:
```diff
@@ -0,0 +1,70 @@
+
+#include <QtCore>
+
+/** \ingroup core
```
I think this should rather go into gui than core (there is no QDesktopServices on server).

elpaso commented on src/core/qgshelp.cpp:
```diff
@@ -0,0 +1,123 @@
+  QUrl helpUrl( url );
+  QTcpSocket socket;
+
+  socket.connectToHost( helpUrl.host(), 80 );
```
Any way to take the Proxy settings into account?

elpaso commented on src/core/qgshelp.cpp:
```diff
@@ -0,0 +1,123 @@
+  {
+    helpPath = QStringLiteral( "%1/%2" ).arg( path ).arg( suffix );
+
+    if ( path.startsWith( QStringLiteral( "http://" ) ) )
```
What if the help system is served through `https` ?
I'd change this check to just startsWith "http" or add an OR condition

m-kuhn commented:
Thanks for starting the work on this!

alexbruy committed:
more robust check for network help location

alexbruy committed:
move QgsHelp to gui

luipir commented:
Sorry, I didn't dig the help problems and solution, but would make sense to have a Help registry as other registry in the QGIS system?
This would need a Help class to be injected using the registry.
The help class would manage the source of the help, so we could have a standard class to get sources as proposed or overlods to get source following a complex logic (in case of offline, or blocked external links).
The help class can manage the indexing and tagging and search optimization.
Would be also a standard mechanism to inject help from plugins.

just my 2c

m-kuhn commented:
@luipir what exactly would be registered in the help registry?

Are you proposing to make help items consist of a `key` and a `source`?
A `source` would normally be `qgis` but plugins could request a help item `key=how_to_turn_water_into_wine source=my_awesome_plugin` and add their own help URL to the documentation paths, and each documentation path can offer one or more `source`s?

I think what we have so far does not have any indexing, it's only context help. A help browser could be done separately in the future (define a standard interface to query a web service or an index file).

alexbruy committed:
add support for proxies

m-kuhn commented:
I actually like the current simple solution with one single widget and a request-based lookup.
IMO we should only start looking for more complex solutions when we are required to do so (e.g. for a search index, but I'm have a feeling that we can solve that as well with the definition of some standard GET/POST or file-based interfaces )

alexbruy commented:
Any objections to merge this?

m-kuhn commented on src/gui/qgshelp.cpp:
```diff
@@ -0,0 +1,169 @@
+    qgisVersion = QStringLiteral( "%1.%2" ).arg( Qgis::QGIS_VERSION_INT / 10000 ).arg( Qgis::QGIS_VERSION_INT / 100 % 100 );
+  }
+
+  QString suffix = QStringLiteral( "%1/%2/docs/user_manual/%3" ).arg( qgisVersion ).arg( qgisLocale ).arg( key );
```
I wonder if the suffix should be always unconditionally prepended to all keys.
Maybe we should instead use an approach with some replacement options in the help search path (e.g. based on qgis variables) and only append the key to this

    http://docs.qgis.org/$qgis_version/$qgis_locale/docs/user_manual/

The main motivation for this would be to offer this system to plugin authors that will not want to structure their docs in the very same way and not related to the qgis version numbering.

alexbruy replied:
Not sure I understand correctly. Original goal was to provide some way to use QGIS docs also as context help/Processing algorithm help source, that's why all keys handled in the same way.

Are you talking about implementing some logic to distinguish QGIS related keys from all other keys and using different help sources for each key type?

m-kuhn replied:
My idea is not to force the path always on each source to be exactly structured like `en/3.0/docs/user_manual/`.

Then I can have a different folder structure locally (ok, not really important), but I can also have my server `https://qfield.org/docs/$qgis_locale` and let my plugin request `QgsHelp.openHelp( 'project-management/portable-project.html' )` to easily use the internal doc viewer also for plugin documentation.

m-kuhn replied:
And in a future change do

    void QgsHelp::openHelp( const QString& key, const QString& source = QString( "qgis" ) );

And allow to register my qfield.org server with a different `source` tag.

alexbruy replied:
Maybe I'm wrong, but it seems that expressions does not allow variable replacement in strings like `https://qfield.org/docs/$qgis_locale`. Instead we should concatenate
`'https://qfield.org/docs/' + $qgis_locale`

m-kuhn replied:
Yes, my proposal was not based on expressions per se. I think expressions are not a very suitable syntax for this task, some simpler syntax would be much easier (based on a simple substitution schema). Not sure if we already have something like this somewhere in QGIS?

alexbruy committed:
[FEATURE] new global expression variable $qgis_locale

Returns current locale used by QGIS. By default this is current
system locale. If user enabled locale override in QGIS settings
overriden locale will be returned.

alexbruy committed:
allow using expression variables in help path. Currently only from
global scope. Syntax should be '$variable_name', e.g. $qgis_version.
When accessing help those variables will be replaced with corresponding
values

alexbruy committed:
add 'qgis_short_version' expression variable to retrieve QGIS version
in the major.minor notation (can be useful for help system and probably
in other places)

alexbruy commented:
@m-kuhn I tried to implement your suggestion about variables substitutions. It is very simple for now and supports only variables from global expressions context scope. Also two new variables were added to make it easier construct help URLs

m-kuhn commented on src/gui/qgshelp.cpp:
```diff
@@ -43,39 +43,25 @@ QUrl QgsHelp::helpUrl( const QString& key )
   Q_FOREACH ( const QString& path, paths )
   {
-    helpPath = QStringLiteral( "%1/%2" ).arg( path ).arg( suffix );
+    qDebug() << "PATH " << path;
```
Is that intended for future use or a leftover from your debugging?
In general `QgsDebugMsg()` should be preferred for messages that are suitable for keeping as debug messages.

m-kuhn commented on src/gui/qgshelp.cpp:
```diff
@@ -43,39 +43,25 @@ QUrl QgsHelp::helpUrl( const QString& key )
+    fullPath = path;
+    Q_FOREACH ( const QString& var, scope->variableNames() )
+    {
+      fullPath.replace( QStringLiteral( "$%1" ).arg( var ), scope->variable( var ).toString() );
```
Can we have an escape syntax here?

`$qgis_version` => `3.2`
vs.
`$$qgis_version` => `$qgis_version`
or similar.

alexbruy committed:
remove debug statements

m-kuhn commented:
Nice, I think that's coming very good! I also think global variables are enough, that's the level at which this configuration works (project / layer would not make any sense)

alexbruy committed:
add missed SIP bindings

m-kuhn commented:
I thought it's easier for a user to remember "$-variable : substitution syntax, @-variable : expression syntax".

alexbruy committed:
support escape syntax. $$qgis_version will be kept and $qgis_version
will be replaced with value

alexbruy commented:
@m-kuhn escape syntax support added. Solution is not perfect, I will think how to improve it.

Any other comments, or we can merge it and start adopting QGIS to use this class?

m-kuhn commented:
Yeah, I think at the moment it will work for `$$some other text $qgis_var` but fail for `$$$qgis_var` (which should be expanded to `$value`).

How about using a regex syntax like the following?

    (\$\$)*(\$%1)?

I'm not a regex expert but I think that will match pairs of `$$` first and then - if there's a remaining `$qgis_var` - it will be replaced.
In a second step, all `$$` can be replaced with a `$`.

Ant then I am very happy to ship this and focus on the server side!

alexbruy committed:
fix regex to correctly handle strings like $$$qgis_version

alexbruy commented:
@m-kuhn should be fixed now

m-kuhn commented:
Sholdn't there be another regex_replace after the variable loop to replace `$$` with `$`?

alexbruy commented:
Sorry, maybe I misunderstand your comment. Will fix it now.

m-kuhn commented:
Actually, there will be still a flaw in my proposal (it's probably not relevant but if you have a better idea, you are welcome). If there's a `$$` inside a variable it will also be escaped to `$` which is not so nice...

alexbruy commented:
Hope now it works fine. Still need to find a way to handle `$$` inside a variable.

m-kuhn commented:
Let's see how this performs in real life :)

I think what we would need to do to get things done properly is

 1. Tokenize to a QStringList `[ "some text", "$$", "$variable", "$ (orphaned $)", etc... ]`
 2. Loop through the tokens, replace as appropriate
 3. Reconstruct string
